Documentation/Yarn-Spec.md

```diff
@@ -0,0 +1,636 @@
+
+#### Operation Output
+
+The following table shows the expected output type of each operation based on its operand type:
+
+|         | +      | -      | *      | /      | %      | ==      | !=      | gt      | lt      | gte     | lte     | \|\|    | &&      | !       | unary minus |
+|---------|--------|--------|--------|--------|--------|---------|---------|---------|---------|---------|---------|---------|---------|---------|-------------|
+| number  | number | number | number | number | number | boolean | boolean | boolean | boolean | boolean | boolean | boolean | boolean |         | number      |
+| string  | string |        |        |        |        | boolean | boolean |         |         |         |         |         |         |         |             |
+| boolean |        |        |        |        |        | boolean | boolean |         |         |         |         | boolean | boolean | boolean |             |
+
+#### Order of Operations
+
+The order of operations is as follows:
+
+1. `()`
+1. `!`, `unary minus`
+1. `*`, `/`, `%`
+1. `-`, `+`
+1. `==`, `!=`, `lt`, `gt`, `lte`, `gte`
+1. `||`, `&&`
+
+If there are any equal priority operations in an expression they are resolved left to right as encountered in the expression.
+
+### Functions
+
+_Functions_ are an alternate way of getting values into expressions.
+Functions are intended to be used to allow more complex code be bundled and called in a different environment, such as in the game itself.
+Functions must return a value.
+
+#### Structure 
+
+Functions are comprised of a name, parentheses, and parameters.
+
+The _function name_ can be any combination of the characters `a` - `z`, `A` - `Z`, and `0` - `9`, but must not start with a number.
+The minimum and maximum length of function names is unspecified.
+
+The parentheses go after the function name and there must be no whitespace between the opening parentheses `(` and the function name.
+The closing parethensis `)` finishes the function.
+
+_Parameters_ go in between the opening `(` and closing `)` parentheses.
+Parameters must be expressions, functions, values, or variables.
+Functions can have multiple parameters, these must be separated by the comma `,` symbol.
+
+Whitespace between parameters and the separator is undefined. hmm this allows for newlines is that ok?
+The maximum and minimum number of allowed parameters a function can have is undefined.
+
+Examples of functions include the following;
+
+```yarn
+getPlayerName()
+DetermineCurrentRoom($playerName, $target, 2)
+rad2Deg(1.5707963268)
+```
+
+#### Handling
+
+The handling of functions by the implementing program is unspecified, however the output type of a function must not change at runtime between calls.
+The implementing program should allow external parts of the game to provide the return value of the function.
+If given the same input parameters multiple invocations of the same functions should return the same value each time.
```

Comment:
Recommend "Care should be taken to appropriately protect against arbitrary code execution made possible by editing function calls in a yarn script."

Reply:
Can you elaborate on that?
Just because currently every function is written by someone else and not something Yarn ever does, it just calls them.

Reply:
If your Yarn script is stored as plaintext in game files then a bad actor (or just someone who's curious) editing it to execute code that the developer didn't intend is trivial. You need to warn people to implement the spec in such a way that they either 1) pass this warning on; 2) provides tools to prevent this (HMAC checksum/encryption/whitelisting); 3) protects against bad actors in some other way